### Heroes Of Pymoli Data Analysis
* Of the 1163 active players, the vast majority are male (84%). There also exists, a smaller, but notable proportion of female players (14%).

* Our peak age demographic falls between 20-24 (44.8%) with secondary groups falling between 15-19 (18.60%) and 25-29 (13.4%).  

### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [1]:
## Dependencies and Setup
import pandas as pd
import numpy as np

## File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)

## Player Count

* Display the total number of players

In [2]:
## total_player is a DataFrame containing just the unique elements SN from purchase_data DataFrame
total_player = pd.DataFrame({"Total Players" : [purchase_data['SN'].nunique()]},index = [0])
total_player

Total Players
0            576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [3]:
## purchasing_analyisis is a DataFrame containing describe info from purchase_data DataFrame
purchasing_analysis = pd.DataFrame(
              {"Number of Unique Items" : [purchase_data['Item ID'].nunique()],
               "Average Price" : [purchase_data['Price'].mean()],
               "Number of Purchases" : [purchase_data['Purchase ID'].nunique()],
               "Total Revenue" : [(purchase_data['Price'].sum())]},
            index = [0])

## Basic formatting for output cells 
purchasing_analysis['Average Price'] = purchasing_analysis['Average Price'].map("${:,.2f}".format)
purchasing_analysis['Total Revenue'] = purchasing_analysis['Total Revenue'].map("${:,.2f}".format)

## Basic formatting for output columns
purchasing_analysis = purchasing_analysis[['Number of Unique Items','Average Price','Number of Purchases','Total Revenue']]
purchasing_analysis


Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     183         $3.05                  780     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [4]:

## gender_analysis is a DataFrame with unique SN from purchase_data DataFrame
gender_analysis = purchase_data.drop_duplicates('SN')

gender_demogra = pd.DataFrame(
              {"Total Count" : [gender_analysis['Gender'].value_counts()[0],
                                gender_analysis['Gender'].value_counts()[1], 
                                gender_analysis['Gender'].value_counts()[2]],
               "Percentage of Players" : [gender_analysis['Gender'].value_counts()[0]*100/len(gender_analysis), 
                                          gender_analysis['Gender'].value_counts()[1]*100/len(gender_analysis),
                                          gender_analysis['Gender'].value_counts()[2]*100/len(gender_analysis)]},
            index = ["Male","Female", "Other / Non-Disclosed"])

## Basic formatting for output cells 
gender_demogra['Percentage of Players'] = gender_demogra['Percentage of Players'].map("{0:.2f}%".format)

## Basic formatting for output columns
gender_demogra = gender_demogra[['Total Count','Percentage of Players']]
gender_demogra



Total Count Percentage of Players
Male                           484                84.03%
Female                          81                14.06%
Other / Non-Disclosed           11                 1.91%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [5]:
## This will provide a quick overview of the data in purchase_data 
purchasing_indicators = purchase_data.groupby('Gender').agg(np.mean) 

## gender_purchasing is a DataFrame containing describe info from purchase_data DataFrame
gender_purchasing = pd.DataFrame(
              {"Purchase Count" : [purchase_data['Gender'].value_counts()[1],
                                purchase_data['Gender'].value_counts()[0], 
                                purchase_data['Gender'].value_counts()[2]],
               "Average Purchase Price" : [purchasing_indicators.iloc[0,3],
                                           purchasing_indicators.iloc[1,3],
                                           purchasing_indicators.iloc[2,3],],
               "Total Purchase Value" : [purchase_data['Gender'].value_counts()[1] * purchasing_indicators.iloc[0,3],
                                         purchase_data['Gender'].value_counts()[0] * purchasing_indicators.iloc[1,3],
                                         purchase_data['Gender'].value_counts()[2] * purchasing_indicators.iloc[2,3],],
               "Avg Total Purchase per Person" : [purchase_data['Gender'].value_counts()[1] * purchasing_indicators.iloc[0,3] / gender_analysis['Gender'].value_counts()[1],
                                                  purchase_data['Gender'].value_counts()[0] * purchasing_indicators.iloc[1,3] / gender_analysis['Gender'].value_counts()[0],
                                                  purchase_data['Gender'].value_counts()[2] * purchasing_indicators.iloc[2,3] / gender_analysis['Gender'].value_counts()[2]]},
               index = ["Female","Male", "Other / Non-Disclosed"])

## Basic formatting for output cells 
gender_purchasing['Average Purchase Price'] = gender_purchasing['Average Purchase Price'].map("${:,.2f}".format)
gender_purchasing['Total Purchase Value'] = gender_purchasing['Total Purchase Value'].map("${:,.2f}".format)
gender_purchasing['Avg Total Purchase per Person'] = gender_purchasing['Avg Total Purchase per Person'].map("${:,.2f}".format)

## Set index for output DataFrame
gender_purchasing.index.name = "Gender"

## Basic formatting for output columns
gender_purchasing = gender_purchasing[['Purchase Count','Average Purchase Price','Total Purchase Value','Avg Total Purchase per Person']]
gender_purchasing




Purchase Count Average Purchase Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Gender                                                                    
Female                             $361.94                         $4.47  
Male                             $1,967.64                         $4.07  
Other / Non-Disclosed               $50.19                         $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [6]:
## Creation of bins an bins-labels to slice gender_analysis data (purchase_data with unique elements)
labels = ['<10','10-14','15-19','20-24','25-29','30-34','35-39','40+']
bins = pd.cut(gender_analysis['Age'], [0, 9, 14, 19, 24, 29, 34, 39, 500],labels=labels)

new_df = gender_analysis.groupby(bins)['Age'].agg(['count'])
new_df['Percentage of Players'] = round(gender_analysis.groupby(bins)['Age'].agg(['count']).apply(lambda x: 100* x / len(gender_analysis)),2)

## Basic formatting for output cells 
new_df['Percentage of Players'] = new_df['Percentage of Players'].map("{0:.2f}%".format)

## Basic formatting for output columns
new_df = new_df.rename(columns={'count' : 'Total Count'})
new_df



Total Count Percentage of Players
Age                                     
<10             17                 2.95%
10-14           22                 3.82%
15-19          107                18.58%
20-24          258                44.79%
25-29           77                13.37%
30-34           52                 9.03%
35-39           31                 5.38%
40+             12                 2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [7]:
## Creation of bins an bins-labels to slice gender_analysis data (purchase_data with unique elements)
labels = ['<10','10-14','15-19','20-24','25-29','30-34','35-39','40+']
bins_purchasing = pd.cut(purchase_data['Age'], [0, 9, 14, 19, 24, 29, 34, 39, 500],labels=labels)

new_purchase_df = purchase_data.groupby(bins_purchasing)['Age'].agg(['count'])
new_purchase_df['Average Purchase Price'] = purchase_data.groupby(bins_purchasing)['Price'].agg(['mean'])
new_purchase_df['Total Purchase Value'] = new_purchase_df['count'] * new_purchase_df['Average Purchase Price']
new_purchase_df['Avg Total Purchase per Person'] = new_purchase_df['Total Purchase Value']/new_df['Total Count']

## Basic formatting for output cells 
new_purchase_df = new_purchase_df.rename(columns={'count' : 'Purchase Count'})

## Basic formatting for output cells 
new_purchase_df['Average Purchase Price'] = new_purchase_df['Average Purchase Price'].map("${:,.2f}".format)
new_purchase_df['Total Purchase Value'] = new_purchase_df['Total Purchase Value'].map("${:,.2f}".format)
new_purchase_df['Avg Total Purchase per Person'] = new_purchase_df['Avg Total Purchase per Person'].map("${:,.2f}".format)
new_purchase_df


Purchase Count Average Purchase Price Total Purchase Value  \
Age                                                                 
<10                23                  $3.35               $77.13   
10-14              28                  $2.96               $82.78   
15-19             136                  $3.04              $412.89   
20-24             365                  $3.05            $1,114.06   
25-29             101                  $2.90              $293.00   
30-34              73                  $2.93              $214.00   
35-39              41                  $3.60              $147.67   
40+                13                  $2.94               $38.24   

      Avg Total Purchase per Person  
Age                                  
<10                           $4.54  
10-14                         $3.76  
15-19                         $3.86  
20-24                         $4.32  
25-29                         $3.81  
30-34                         $4.12  
35-39                         $4.76  
40+                           $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [8]:
## Top_Spenders SN and Total Purchase Value
# Group by SN and sum price and give me just top 5 
top_spenders_analysis = purchase_data.groupby(by=['SN'])['Price'].sum().nlargest(5)

# Add a DataSeries with top_spender_analysis top 5 values 
data = ({'Total Purchase Value': top_spenders_analysis})
# Add data info to a Pandas DataFrame called top_purch_sn_df
top_purch_sn_df = pd.DataFrame(data,columns=['Total Purchase Value'])
top_purch_sn_df

# Group by SN and count how many purchases ID are in each SN
data_purchase_count = purchase_data.groupby(by=['SN'])['Purchase ID'].count()
# Add data_purchase_count info to a Pandas DataFrame called top_purch_count_df
top_purch_count_df = pd.DataFrame({'Purchase Count':data_purchase_count},columns=['Purchase Count'])
top_purch_count_df

# Add Purchase Count column to top_purch_sn_df with top_purch_count_df info 
top_purch_sn_df['Purchase Count'] = top_purch_count_df['Purchase Count']

# Calculate info related to Avg
top_purch_sn_df['Average Purchase Price'] = top_purch_sn_df['Total Purchase Value'] / top_purch_sn_df['Purchase Count']


## Basic formatting for output cells 
top_purch_sn_df['Average Purchase Price'] = top_purch_sn_df['Average Purchase Price'].map("${:,.2f}".format)
top_purch_sn_df['Total Purchase Value'] = top_purch_sn_df['Total Purchase Value'].map("${:,.2f}".format)

top_purch_sn_df = top_purch_sn_df[['Purchase Count','Average Purchase Price','Total Purchase Value']]
top_purch_sn_df

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [10]:

##TOP NAMES DATAFRAME
## To get top names on Purchase ID count() grouped by Item ID
purchase_data['names_top'] = purchase_data.groupby(by=['Item ID'])['Purchase ID'].count()
## To get the new DataFrame sorted by names_top higher values
top_names_df = purchase_data.sort_values(by=['names_top'], ascending=False)

## Reduce top_names_df for better undertstanding
top_names_df = top_names_df[['Purchase ID','Item ID','Item Name']]

## Rename column(s)
top_names_df = top_names_df.rename(columns={'Item Name' : 'Nombre nuevo'})


##TOP ITEMS DATAFRAME
## Add column "top_items" to purchase data and group info by Item_ID and count the 'Purchase ID', top 5
purchase_data['top_items'] = purchase_data.groupby(by=['Item ID'])['Purchase ID'].count().nlargest(5)
top_items_df = purchase_data.sort_values(by=['top_items'], ascending=False).nlargest(5, 'top_items')
top_items_df = top_items_df.sort_values(by=['Purchase ID'], ascending=False)


##TOP VALUES DATAFRAME
## Add column "top_values" to purchase data and group info by Item_ID and sum total value 'Price'
purchase_data['top_values'] = purchase_data.groupby(by=['Item ID'])['Price'].sum()
## To get the new DataFrame sorted by top_values values
top_values_df = purchase_data.sort_values(by=['top_values'], ascending=False)
top_values_df

# Calculate Item Price value based on top_values/top_items
top_values_df['Item Price'] = top_values_df['top_values'] / top_values_df['top_items']
top_values_df
# ## Merge data from top_items_df & top_values_df
top_items_df = pd.merge(top_items_df, top_values_df, on='Purchase ID')
top_items_df

## To get Item price 

## Rename column(s)
top_items_df = top_items_df.rename(columns={'Purchase ID' : 'Item ID'})
top_items_df = top_items_df.rename(columns={'top_items_y' : 'Purchase Count'})
top_items_df = top_items_df.rename(columns={'top_values_y' : 'Total Purchase Value'})

## Basic formatting for currency and counts
top_items_df['Purchase Count'] = top_items_df['Purchase Count'].map("{:,.0f}".format)
top_items_df['Item Price'] = top_items_df['Item Price'].map("${:,.2f}".format)
top_items_df['Total Purchase Value'] = top_items_df['Total Purchase Value'].map("${:,.2f}".format)

## Merge data from top_items_df & top_names_df
top_items_df = pd.merge(top_items_df,top_names_df, on='Item ID') 
top_items_df


## Arrange final columns for the output
most_popular_df = top_items_df[['Item ID','Nombre nuevo','Purchase Count','Item Price', 'Total Purchase Value']]

## Drop duplicates for Nombre nuevo 
most_popular_df = most_popular_df.drop_duplicates('Nombre nuevo')

## Rename column(s)
most_popular_df  = most_popular_df.rename(columns={'Nombre nuevo' : 'Item Name'})

## Set multiple index(s) for most_popular_df
# most_popular_df.set_index(['Item ID','Nombre nuevo'])
most_popular_df.set_index(['Item ID','Item Name'])




Purchase Count  \
Item ID Item Name                                                     
178     Oathbreaker, Last Hope of the Breaking Storm             12   
145     Fiery Glass Crusader                                      9   
108     Extraction, Quickblade Of Trembling Hands                 9   
82      Nirvana                                                   9   
19      Pursuit, Cudgel of Necromancy                             8   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
108     Extraction, Quickblade Of Trembling Hands         $3.53   
82      Nirvana                                           $4.90   
19      Pursuit, Cudgel of Necromancy                     $1.02   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
108     Extraction, Quickblade Of Trembling Hands                  $31.77  
82      Nirvana                                                    $44.10  
19      Pursuit, Cudgel of Necromancy                               $8.16

In [11]:
# Group purchase_data DataFrame by Item ID and top 5 
purchase_data['top_profit_values'] = purchase_data.groupby(by=['Item ID'])['Price'].sum()
top_profit_values_df = purchase_data.sort_values(by=['Item ID'], ascending=False).nlargest(5, 'top_profit_values')

# Filter columns
profit_top_names_df = top_names_df[['Item ID', 'Nombre nuevo']]

# Rename Item ID to Purchase ID 
profit_top_names_df = profit_top_names_df.rename(columns={'Item ID' : 'Purchase ID'})

#merge top_profit_values_df with top_names_df
top_profit_values_df = pd.merge(top_profit_values_df,profit_top_names_df, on = 'Purchase ID')

# Drop duplicates
top_profit_values_df = top_profit_values_df.drop_duplicates('Nombre nuevo')

# Caluclate Item Price 
top_profit_values_df['Item Price'] = top_profit_values_df['top_values'] / top_profit_values_df['names_top']


# top_profit_values_df = top_profit_values_df[['Purchase ID','top_values','Nombre nuevo', 'names_top','Item Price']]

top_profit_values_df = top_profit_values_df[['Purchase ID','Nombre nuevo','names_top','Item Price','top_values']]


## Rename column(s)
top_profit_values_df = top_profit_values_df.rename(columns={'Purchase ID' : 'Item ID'})
top_profit_values_df = top_profit_values_df.rename(columns={'Nombre nuevo' : 'Item Name'})
top_profit_values_df = top_profit_values_df.rename(columns={'names_top' : 'Purchase Count'})
top_profit_values_df = top_profit_values_df.rename(columns={'top_values' : 'Total Purchase Value'})

## Basic formatting for currency and counts
top_profit_values_df['Purchase Count'] = top_profit_values_df['Purchase Count'].map("{:,.0f}".format)
top_profit_values_df['Item Price'] = top_profit_values_df['Item Price'].map("${:,.2f}".format)
top_profit_values_df['Total Purchase Value'] = top_profit_values_df['Total Purchase Value'].map("${:,.2f}".format)
top_profit_values_df

top_profit_values_df.set_index(['Item ID','Item Name'])

Purchase Count  \
Item ID Item Name                                                     
178     Oathbreaker, Last Hope of the Breaking Storm             12   
82      Nirvana                                                   9   
145     Fiery Glass Crusader                                      9   
92      Final Critic                                              8   
103     Singed Scalpel                                            8   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
92      Final Critic                                      $4.88   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
92      Final Critic                                               $39.04  
103     Singed Scalpel                                             $34.80

Conclusions for Heroes of Pymoli by Daniel Cespedes

1. The vast majority of the players (44.79%) are between ages 20-24, representing a total of 258 players. 

2. The major influencer for the amount of Total Purchase Value are the group ofplayers between ages 20-24. 
                                    
3. Our golden user who is the top spender is the SN Lisosia93 with these amounts: 
Purchase Count: 5	Average Purchase Price: $3.79	Total Purchase Value: $18.96			
    